app/server.js

```diff
@@ -1,31 +1,212 @@
+  const validatedUser = {
+    name: user.name,
+    email: user.email,
+    token: uid(16),
+    lastUpdated: new Date(),
+  };
+  validatedUsers.push(validatedUser);
+
+  return response.status(200).json(validatedUser);
+});
+
+app.get("/api/sunglasses/brands", (request, response) => {
+  const brand = request.query?.brand;
+
+  if (!brand) {
+    return response
+      .status(400)
+      .json({ message: "Brand name required in query" });
+  }
+  if (brand) {
+    const validBrandFromQuery = brands.find(
+      (brandForSearch) => brandForSearch.name === brand
+    );
+
+    if (!validBrandFromQuery) {
+      return response.status(404).json({ message: "Brand not found" });
+    }
+
+    const categoryId = brands.find(
+      (brandForSearch) => brandForSearch.name === brand
+    ).id;
+
+    const sunglasses = products.filter(
+      (product) => product.categoryId === categoryId
+    );
+
+    return response.status(200).json(sunglasses);
+  }
+});
+
+app.get("/api/sunglasses/search", (request, response) => {
+  const search = request.query?.search?.toLowerCase();
+
+  if (!search) {
+    return response.status(400).json({ message: "Search query required" });
+  }
+
+  if (search) {
+    const sunglasses = products.filter(
+      (product) =>
+        product.name.toLowerCase().includes(search) ||
+        product.description.toLowerCase().includes(search)
+    );
+
+    if (sunglasses.length === 0) {
+      return response
+        .status(404)
+        .json({ message: "No sunglasses match your search" });
+    }
+
+    return response.status(200).json(sunglasses);
+  }
+});
+
+app.post("/api/me/cart/:itemId", (request, response) => {
+  const product = products.find(
+    (product) => product.id === request.params.itemId
+  );
+
+  if (!product) {
+    return response.status(404).json({ message: "Invalid product id" });
+  }
+
+  const { currentValidatedUser, isValidToken } = authenticate(request);
+  if (!isValidToken) {
+    return response.status(401).json({
+      message: "Login requried to add items to cart",
+    });
+  }
+
+  for (let i = 0; i < users.length; i++) {
+    if (users[i].email === currentValidatedUser.email) {
+      users[i].cart.push(product);
+      break;
+    }
+  }
+  return response
+    .status(200)
+    .json({ message: `Item ${product.id} added to cart` });
+});
+
+app.delete("/api/me/cart/:itemId", (request, response) => {
+  const idToRemove = request.params.itemId;
+
+  const product = products.find((product) => product.id === idToRemove);
+
+  if (!product) {
+    return response.status(404).json({ message: "Invalid product id" });
+  }
+
+  const { currentValidatedUser, isValidToken } = authenticate(request);
+
+  if (!isValidToken) {
+    return response.status(401).json({
+      message: "Login requried to delete items from cart",
+    });
+  }
+
+  for (let i = 0; i < users.length; i++) {
+    if (users[i].email === currentValidatedUser.email) {
+      const productIndex = users[i].cart.findIndex(
```

Comment:
why not users[i].cart.find ?

Reply:
nesting loops is ok for this level but its not maintainable for larger datasets of either users of products. This will grow exponentially the iterations. This area is a very important part of the interview questions. 
Check suggestion: